# Day 1

In [1]:
import io
import zipfile
import requests
import frontmatter

def read_repo_data(repo_owner, repo_name):
    """
    Download and parse all markdown files from a GitHub repository.
    
    Args:
        repo_owner: GitHub username or organization
        repo_name: Repository name
    
    Returns:
        List of dictionaries containing file content and metadata
    """
    prefix = 'https://codeload.github.com' 
    url = f'{prefix}/{repo_owner}/{repo_name}/zip/refs/heads/main'
    resp = requests.get(url)
    
    if resp.status_code != 200:
        raise Exception(f"Failed to download repository: {resp.status_code}")

    repository_data = []
    zf = zipfile.ZipFile(io.BytesIO(resp.content))
    
    for file_info in zf.infolist():
        filename = file_info.filename
        filename_lower = filename.lower()

        if not (filename_lower.endswith('.md') 
            or filename_lower.endswith('.mdx')):
            continue
    
        try:
            with zf.open(file_info) as f_in:
                content = f_in.read().decode('utf-8', errors='ignore')
                post = frontmatter.loads(content)
                data = post.to_dict()
                data['filename'] = filename
                repository_data.append(data)
        except Exception as e:
            print(f"Error processing {filename}: {e}")
            continue
    
    zf.close()
    return repository_data

In [2]:
dtc_faq = read_repo_data('DataTalksClub', 'faq')
evidently_docs = read_repo_data('evidentlyai', 'docs')

print(f"FAQ documents: {len(dtc_faq)}")
print(f"Evidently documents: {len(evidently_docs)}")

FAQ documents: 1285
Evidently documents: 95


In [3]:
learning_data_engineering_docs = read_repo_data('Joyan9', 'learning_data_engineering')
print(f"LDE documents: {len(learning_data_engineering_docs)}")

LDE documents: 47


# Day 2

In [4]:
len(evidently_docs[45]['content'])

21712

### Sliding Window Chunking

The content field is 21,712 characters long. The simplest thing we can do is cut it into pieces of equal length. For example, for size of 2000 characters

However, this approach has disadvantages:

- Context loss: Important information might be split in the middle
- Incomplete sentences: Chunks might end mid-sentence
- Missing connections: Related information might end up in different chunks

That's why, in practice, we usually make sure there's overlap between chunks. For size 2000 and overlap 1000, we will have:

- Chunk 1: 0..2000
- Chunk 2: 1000..3000
- Chunk 3: 2000..4000

**This approach is called the sliding window chunking**

In [3]:
def sliding_window(seq, size, step):
    if size <= 0 or step <= 0:
        raise ValueError("size and step must be positive")

    n = len(seq)
    result = []
    for i in range(0, n, step):
        chunk = seq[i:i+size]
        result.append({'start': i, 'chunk': chunk})
        if i + size >= n:
            break

    return result

In [4]:
evidently_chunks = []

for doc in evidently_docs:
    # create a shollow copy of each doc
    doc_copy = doc.copy()
    # the pop method removes the content key and returns the value
    doc_content = doc_copy.pop('content')
    chunks = sliding_window(doc_content, size=2000, step=1000)
    for chunk in chunks:
        chunk.update(doc_copy)
    evidently_chunks.extend(chunks)


In [5]:
len(evidently_chunks)

576

In [6]:
# splitting by paragraphs and sections

import re

def split_markdown_by_level(text, level=2):
    """
    Split markdown text by a specific header level.
    
    :param text: Markdown text as a string
    :param level: Header level to split on
    :return: List of sections as strings
    """
    # This regex matches markdown headers
    # For level 2, it matches lines starting with "## "
    header_pattern = r'^(#{' + str(level) + r'} )(.+)$'
    pattern = re.compile(header_pattern, re.MULTILINE)

    # Split and keep the headers
    parts = pattern.split(text)
    
    sections = []
    for i in range(1, len(parts), 3):
        # We step by 3 because regex.split() with
        # capturing groups returns:
        # [before_match, group1, group2, after_match, ...]
        # here group1 is "## ", group2 is the header text
        header = parts[i] + parts[i+1]  # "## " + "Title"
        header = header.strip()

        # Get the content after this header
        content = ""
        if i+2 < len(parts):
            content = parts[i+2].strip()

        if content:
            section = f'{header}\n\n{content}'
        else:
            section = header
        sections.append(section)
    
    return sections


In [7]:
evidently_chunks_split_by_level = []

for doc in evidently_docs:
    doc_copy = doc.copy()
    doc_content = doc_copy.pop('content')
    sections = split_markdown_by_level(doc_content, level=2)
    for section in sections:
        section_doc = doc_copy.copy()
        section_doc['section'] = section
        evidently_chunks_split_by_level.append(section_doc)

len(evidently_chunks_split_by_level)

266

In [8]:
import os

from groq import Groq

client = Groq(
    # This is the default and can be omitted
    api_key=os.environ.get("GROQ_API_KEY"),
)

chat_completion = client.chat.completions.create(
    messages=[
        {
            "role": "system",
            "content": "You are a helpful assistant."
        },
        {
            "role": "user",
            "content": "Explain the importance of fast language models",
        }
    ],
    model="llama-3.3-70b-versatile",
)

print(chat_completion.choices[0].message.content)

Fast language models are crucial in today's technology landscape, and their importance can be attributed to several factors:

1. **Efficient Processing**: Fast language models can process and analyze vast amounts of text data quickly, making them ideal for applications where speed is essential, such as real-time chatbots, virtual assistants, and language translation systems.
2. **Improved User Experience**: By providing rapid responses to user queries, fast language models can significantly enhance the user experience. This is particularly important in customer service, where timely responses can lead to increased customer satisfaction and loyalty.
3. **Increased Productivity**: Fast language models can automate many tasks, such as text summarization, sentiment analysis, and content generation, allowing humans to focus on higher-level tasks that require creativity, critical thinking, and problem-solving skills.
4. **Competitive Advantage**: Organizations that leverage fast language mod

In [13]:
# chunking with a LLM - useful when the documentation is not structured, the simpler methods do not provide adequate results, and you have a budget

prompt_template = """
Split the provided document into logical sections
that make sense for a Q&A system.

Each section should be self-contained and cover
a specific topic or concept.

<DOCUMENT>
{document}
</DOCUMENT>

Use this format:

## Section Name

Section content with all relevant details

---

## Another Section Name

Another section content

---
""".strip()

import os

from groq import Groq

client = Groq(
    api_key=os.environ.get("GROQ_API_KEY"),
)


def llm(prompt, model="meta-llama/llama-4-scout-17b-16e-instruct"):

    chat_completion = client.chat.completions.create(messages=[
                                                                {
                                                                    "role": "system",
                                                                    "content": "You are a helpful assistant."
                                                                },
                                                                {
                                                                    "role": "user",
                                                                    "content": prompt,
                                                                }
                                                                ],
                                                                model= model,
                                                                )

    output_text = chat_completion.choices[0].message.content

    return output_text
    

def intelligent_chunking(text):
    prompt = prompt_template.format(document=text)
    response = llm(prompt)
    sections = response.split('---')
    sections = [s.strip() for s in sections if s.strip()]
    return sections

In [14]:
from tqdm.auto import tqdm
# tqdm is a library that shows progress bars. It helps you track progress when processing a large number of documents.

evidently_chunks_split_by_llm = []

# processing only the first 45 documents to avoid rate limit error
for doc in tqdm(evidently_docs[:45]):
    doc_copy = doc.copy()
    doc_content = doc_copy.pop('content')

    sections = intelligent_chunking(doc_content)
    for section in sections:
        section_doc = doc_copy.copy()
        section_doc['section'] = section
        evidently_chunks_split_by_llm.append(section_doc)

  0%|          | 0/45 [00:00<?, ?it/s]

In [16]:
evidently_chunks_split_by_llm_2 = []

for doc in tqdm(evidently_docs[45:]):
    doc_copy = doc.copy()
    doc_content = doc_copy.pop('content')

    sections = intelligent_chunking(doc_content)
    for section in sections:
        section_doc = doc_copy.copy()
        section_doc['section'] = section
        evidently_chunks_split_by_llm_2.append(section_doc)

  0%|          | 0/50 [00:00<?, ?it/s]

In [15]:
len(evidently_chunks_split_by_llm)

155

In [17]:
len(evidently_chunks_split_by_llm_2)

183